Suitor matcher

angriman opened the pull request:
This PR brings the 1/2-approximation algorithm for maximum (weighted) matching by Manne and Halappanavar (New Effective Multithreaded Matching Algorithms, IPDPS 2014).
There are two versions of the algorithm, one of which only supports graphs where the adjacency lists of every vertex are sorted by edge weight in decreasing order. To this end I also added a new `sortEdgesByWeight` method in `Graph`, which allows to sort the edges of a graph by edge weight in increasing or decreasing order. This method uses a new generic `sortEdges` method in `Graph`, which allows to sort the edges of a graph according to a custom criterion.

I am a bit hesitant to add the new sorting methods to `Graph`, they are probably too specific. An alternative idea is to implement them in `GraphTools`.

avdgrinten commented:
Does the increase in runtime performance really justify the templated implementation? Can't you just branch on a sortSuitor variable at runtime?

Also, the sortEdgesByWeight method is not appropriate for `Graph`, as it can be implemented based on simpler primitives (i.e., the sortEdges method).

angriman commented:
> Does the increase in runtime performance really justify the templated implementation? Can't you just branch on a sortSuitor variable at runtime?

I removed the templated implementation, it that was not really necessary.

> Also, the sortEdgesByWeight method is not appropriate for `Graph`, as it can be implemented based on simpler primitives (i.e., the sortEdges method).

I moved it in `GraphTools` and it uses the `sortEdges` primitive.

michitux commented on include/networkit/graph/Graph.hpp:
```diff
@@ -2162,6 +2186,57 @@ std::pair<count, count> Graph::removeAdjacentEdges(node u, Condition condition,
+                          WeightedEdge{u, adjList[b], weights[b]});
+        });
+
+        std::vector<node> sortedAdjList(deg);
+        std::vector<edgeweight> sortedWeights(deg);
+        for (index i = 0; i < deg; ++i) {
+            sortedAdjList[i] = adjList[indices[i]];
+            sortedWeights[i] = weights[indices[i]];
+        }
+
+        adjList = std::move(sortedAdjList);
+        weights = std::move(sortedWeights);
```
Wouldn't it be possible to apply this permutation in-place using just a single temporary value? This could also be done in a utility method that is then applied to the vectors as needed.

angriman replied:
That's a good idea, I implemented this feature in `ArrayTools::applyPermutation`. Unfortunately I a have to use a bool vector to keep track of which elements have already been swapped, not sure if we can avoid that.

michitux replied:
It is not easily possible to avoid this, the best solution seems to use a bit of the permutation to mark the swapped elements (e.g., used signed integers and change each value `x` to `- 1 - x`). Afterwards, one would need to reverse this.

angriman replied:
This would work if the permutation does not use all the bits right? If it does, we still need to use an additional vector. I could add a condition that checks if the length of the permutation requires all the bits, and use the additional vector only if necessary.

michitux replied:
If the permutation needs all 64 bits, this means that the graph has more than 2^63 nodes. I do not think that NetworKit intends to support such graphs in the foreseeable future.

angriman replied:
So do we always assume that the permutation is a 64-bit integer? Since this is a general utility method it might make sense to support also shorter integers (e.g., (u)int8_t).

michitux commented on networkit/graph.pyx:
```diff
@@ -394,6 +394,15 @@ cdef class Graph:
+				self._this.sortEdges[SortEdgesLambdaWrapper](dereference(wrapper))
+		finally:
+			del wrapper
+
```
Is there any test for this method? As this may be a bit fragile I would suggest to add a test case for it.

angriman replied:
This was a first attempt to provide `sortEdgesByWeight` to Python, which is now implemented in `nk.graphtools`. Since it is not needed anymore I think it is a better idea to keep it for C++ only.

michitux commented on networkit/cpp/matching/SuitorMatcher.cpp:
```diff
@@ -0,0 +1,148 @@
+}
+
+bool SuitorMatcher::hasEdgesSortedByWeight(const Graph &G) {
+    for (node u : G.nodeRange()) {
```
This could be parallelized easily but on the other hand it might not be worth the effort.

michitux commented on networkit/cpp/matching/SuitorMatcher.cpp:
```diff
@@ -0,0 +1,148 @@
+                sorted = false;
+
+            prevNode = v;
+            prevWeight = ew;
```
This seems like a re-implementation of [`std::is_sorted`](https://en.cppreference.com/w/cpp/algorithm/is_sorted) on `G.weightNeighborRange` with a custom comparison function.

michitux commented on networkit/cpp/matching/SuitorMatcher.cpp:
```diff
@@ -0,0 +1,148 @@
+
+        for (index &iter = neighborIterators[current]; iter < G->degree(current); ++iter) {
+            const node v = G->getIthNeighbor(current, iter);
+            const edgeweight weight = G->getIthNeighborWeight(current, iter);
```
Why does `neighborIterators` not actually store an iterator (e.g., a `NeighborWeightRange`)?

angriman replied:
If the input graph is unweighted then `NeighborWeightRange` cannot be used because it requires an unweighted graph. To make this work we need to use two vectors with two different iterators and add some checks to pick the correct one at runtime, but I find it easier to just store the index to the current neighbor.
However, I think it would be a better idea to make `NeighborWeightRange` also work with unweighted graphs, that would make things easier.

hmeyerhenke replied:
You mean "...because it requires a weighted graph", right?

angriman replied:
Yes, thanks.

michitux commented on include/networkit/graph/Graph.hpp:
```diff
@@ -1535,6 +1556,20 @@ class Graph final {
+     * @return @a edge weight to the i-th (outgoing) neighbor of @a u, or @c +inf if no such
+     * neighbor exists.
+     */
+    edgeweight getIthNeighborWeight(node u, index i) const {
```
Is it intentional that this method is not added to the Python interface?

angriman replied:
`getIthNeighbor` does not have a Python interface, so we either provide a Python interface for both or for none.

michitux commented on include/networkit/graph/Graph.hpp:
```diff
@@ -54,6 +56,16 @@ struct WeightedEdge : Edge {
+struct WeightedEdgeWithId : WeightedEdge {
+    edgeid eid;
+
+    // Needed by cython
```
There seems to be no Cython interface for this currently, so this comment seems misleading.

michitux reviewed (approved):
In general this looks good now, I have just a few small comments that you may implement or not.

Do you have any idea how fast sorting neighbors by id with your new function is compared to the existing `sortEdges` functionality? My suspicion would be that due to increased locality it may be faster in practice, in particular when multiple cores are used. If it should be faster, I would suggest to replace the existing `sortEdges` method by your new method.

fabratu commented on include/networkit/graph/Graph.hpp:
```diff
@@ -2162,6 +2197,61 @@ std::pair<count, count> Graph::removeAdjacentEdges(node u, Condition condition,
+    const auto sortAdjacencyArrays = [&](node u, std::vector<node> &adjList,
+                                         std::vector<edgeweight> &weights,
+                                         std::vector<edgeid> &edgeIds) -> void {
+        std::vector<index> indices(adjList.size());
```
Although the code slightly changed in the most recent version, I think time can still be safed if resize is only done if `adjList.size() > indices.size()`. Reasoning: resize calls erase on indices if `adjList.size() < indices.size()`, while iota is called anyways afterwards.

angriman replied:
Now resize is called only when necessary.

fabratu reviewed (approved):
LGTM.
